math/dec_test.go

```diff
@@ -1,1028 +1,479 @@
 	}
-	data := make([]byte, 100)
-
-	b.ReportAllocs()
-	b.ResetTimer()
-
-	for i := 0; i < b.N; i++ {
-		for _, bi := range bis {
-			if n, err := bi.in.MarshalTo(data); err != nil {
-				b.Fatal(err)
-			} else if !bytes.Equal(data[:n], bi.want) {
-				b.Fatalf("Mismatch\nGot:  % x\nWant: % x\n", data[:n], bi.want)
-			}
-		}
+	for name, spec := range specs {
+		t.Run(name, func(t *testing.T) {
+			got := NewDecFromInt64(spec.src)
+			require.NoError(t, spec.expErr)
+			assert.Equal(t, spec.exp, got.String())
+		})
 	}
 }
 
-var sink interface{}
-
-func BenchmarkLegacyQuoMut(b *testing.B) {
-	b1 := math.LegacyNewDec(17e2 + 8371)
-	b2 := math.LegacyNewDec(4371)
-	b.ReportAllocs()
-	b.ResetTimer()
-	for i := 0; i < b.N; i++ {
-		sink = b1.QuoMut(b2)
+func TestAdd(t *testing.T) {
+	specs := map[string]struct {
+		x           Dec
+		y           Dec
+		constraints []SetupConstraint
+		exp         Dec
+		expErr      error
+	}{
+		"zero add zero": {
+			// 0 + 0 = 0
+			x:   NewDecFromInt64(0),
+			y:   NewDecFromInt64(0),
+			exp: NewDecFromInt64(0),
+		},
+		"zero add simple positive": {
+			// 0 + 123 = 123
+			x:   NewDecFromInt64(0),
+			y:   NewDecFromInt64(123),
+			exp: NewDecFromInt64(123),
+		},
+		"zero and simple negative": {
+			// 0 + -123 = -123
+			x:   NewDecFromInt64(0),
+			y:   NewDecFromInt64(-123),
+			exp: NewDecFromInt64(-123),
+		},
+		"simple positive add simple positive": {
+			// 123 + 123 = 246
+			x:   NewDecFromInt64(123),
+			y:   NewDecFromInt64(123),
+			exp: NewDecFromInt64(246),
+		},
+		"simple negative add simple positive": {
+			// -123 + 123 = 0
+			x:   NewDecFromInt64(-123),
+			y:   NewDecFromInt64(123),
+			exp: NewDecFromInt64(0),
+		},
+		"simple negative add simple negative": {
+			// -123 + -123 = -246
+			x:   NewDecFromInt64(-123),
+			y:   NewDecFromInt64(-123),
+			exp: NewDecFromInt64(-246),
+		},
+		"valid decimal with decimal places add valid decimal with decimal places": {
+			// 1.234 + 1.234 = 2.468
+			x:   NewDecWithPrec(1234, -3),
+			y:   NewDecWithPrec(1234, -3),
+			exp: NewDecWithPrec(2468, -3),
+		},
+		"valid decimal with decimal places and simple positive": {
+			// 1.234 + 123 = 124.234
+			x:   NewDecWithPrec(1234, -3),
+			y:   NewDecFromInt64(123),
+			exp: NewDecWithPrec(124234, -3),
+		},
+		"valid decimal with decimal places and simple negative": {
+			// 1.234 + -123 = 1.111
+			x:   NewDecWithPrec(1234, -3),
+			y:   NewDecFromInt64(-123),
+			exp: NewDecWithPrec(111, -3),
+		},
+
+		"valid decimal with decimal places add valid negative decimal with decimal places": {
+			// 1.234 + -1.234 = 0
+			x:   NewDecWithPrec(1234, -3),
+			y:   NewDecWithPrec(-1234, -3),
+			exp: NewDecWithPrec(0, -3),
+		},
+		"valid negative decimal with decimal places add valid negative decimal with decimal places": {
+			// -1.234 + -1.234 = -2.468
+			x:   NewDecWithPrec(-1234, -3),
+			y:   NewDecWithPrec(-1234, -3),
+			exp: NewDecWithPrec(-2468, -3),
+		},
+		// "precision too high": {
```

Comment:
good set of tests already. I would love to see some cases that include roundings before and after the comma.
Also the edge cases on the max/min exponents get really interesting.  For example: `1^100000 + 9^100000` , `1^100000 + 1^-1`,...